keps/sig-api-machinery/5116-streaming-response-encoding/README.md

```diff
@@ -0,0 +1,357 @@
+* Handling structs implementing `MarshallJSON` method, especially built-in collection types.
+* Handling raw bytes.
+* Linting unit test to ensure all our built-in collection types would be matched.
+
+Fuzz tests will cover the custom streaming encoders for the types with overwritten encoders:
+* `testingapigroup.CarpList` as surrogate for built-in types
+* `UnstructuredList`
+
+The skipped tests are primarily related to decoding or CBOR-specific features, which are not relevant to the streaming encoding of JSON and Proto addressed by this KEP. 
+
+##### Integration tests
+
+With one to one compatibility to the existing encoder we don't expect integration tests between components will be needed.
+
+##### e2e tests
+
+Scalability tests that will confirm the improvements and protect against future regressions. 
+Improvements in the resources should be noticiable in on the perf-dash. 
+
+The tests will cover the following properties:
+* Large resource, 10000 objects each 100KB size.
+* List with `RV=0` to ensure response is served from watch cache and all the overhead comes from encoder memory allocation.
+* Different content type JSON (default), Proto, CBOR, YAML.
+* Different API kinds, eg ConfigMap, Pod, custom resources
+
+In first iteration we expect we will overallocate the resources needed for apiserver to ensure passage, 
+however after the improvement is implemented we will tune down the resources to detect regressions.
+
+### Graduation Criteria
+
+#### Beta
+
+- Gzip compression is supporting chunking
+- All encoder unit tests are implemented
+- Streaming encoder for JSON and Proto are implemented
+- Scalability test are running and show improvement
+
+#### GA
+
+- Scalability tests are release blocking
+
+
+### Upgrade / Downgrade Strategy
+
+We plan to provide byte to byte compatibility.
+
+### Version Skew Strategy
+
+We plan to provide byte to byte compatibility.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+Via feature gates
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: StreamingCollectionEncodingToJSON, StreamingCollectionEncodingToProto
+  - Components depending on the feature gate: kube-apiserver
+
+###### Does enabling the feature change any default behavior?
+
+No, we provide byte to byte compatibility.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, without problem.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, will be covered by unit tests.
+
+### Rollout, Upgrade and Rollback Planning
+
+N/A
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+N/A
+
+###### What specific metrics should inform a rollback?
+
+N/A
```

Comment:
this doesn't seem correct.  Latency and failed list request metrics should be checked.

Reply:
Done